pkg/operator/ceph/cluster/osd/create.go

```diff
@@ -0,0 +1,391 @@
+)
+
+func (c *Cluster) newCreateConfig(
+	provisionConfig *provisionConfig,
+	awaitingStatusConfigMaps *util.Set,
+	deployments *existenceList,
+) *createConfig {
+	if awaitingStatusConfigMaps == nil {
+		awaitingStatusConfigMaps = util.NewSet()
+	}
+	return &createConfig{
+		c,
+		provisionConfig,
+		awaitingStatusConfigMaps,
+		awaitingStatusConfigMaps.Count(),
+		deployments,
+	}
+}
+
+func (c *createConfig) progress() (completed, initial int) {
+	return (c.numJobsCreated - c.awaitingStatusConfigMaps.Count()), c.numJobsCreated
+}
+
+func (c *createConfig) doneCreating() bool {
+	return c.awaitingStatusConfigMaps.Count() == 0
+}
+
+func (c *createConfig) createNewOSDsFromStatus(
+	status *OrchestrationStatus,
+	nodeOrPVCName string,
+	errs *provisionErrors,
+) {
+	if !c.awaitingStatusConfigMaps.Contains(statusConfigMapName(nodeOrPVCName)) {
+		// If there is a dangling OSD prepare configmap from another reconcile, don't process it
+		logger.Infof("not creating deployments for OSD prepare results found in ConfigMap %q which was not created for the latest storage spec", statusConfigMapName(nodeOrPVCName))
+		return
+	}
+	for _, osd := range status.OSDs {
+		if c.deployments.Exists(osd.ID) {
+			// This OSD will be handled by the updater
+			logger.Debugf("not creating deployment for OSD %d which already exists", osd.ID)
+			continue
+		}
+		if status.PvcBackedOSD {
+			logger.Infof("creating OSD %d on PVC %q", osd.ID, nodeOrPVCName)
+			err := createOSDDaemonOnPVC(c.cluster, osd, nodeOrPVCName, c.provisionConfig)
+			if err != nil {
+				errs.addError("%v", errors.Wrapf(err, "failed to create OSD %d on PVC %q", osd.ID, nodeOrPVCName))
+			}
+		} else {
+			logger.Infof("creating OSD %d on node %q", osd.ID, nodeOrPVCName)
+			err := createOSDDaemonOnNode(c.cluster, osd, nodeOrPVCName, c.provisionConfig)
+			if err != nil {
+				errs.addError("%v", errors.Wrapf(err, "failed to create OSD %d on node %q", osd.ID, nodeOrPVCName))
+			}
+		}
+	}
+
+	// indicate progress processing prepare jobs by removing the status configmap from the list
+	c.awaitingStatusConfigMaps.Remove(statusConfigMapName(nodeOrPVCName))
+}
+
+// NOTE: moved this to create.go because creation of prepare jobs is most directly related to
+// creation of new OSDs. We want to keep all usage of awaitingStatusConfigMaps in this file.
+// Returns error only if the calling function should halt all OSD provisioning. Non-halting errors
+// are added to provisionErrors.
+func (c *Cluster) startProvisioningOverPVCs(config *provisionConfig, errs *provisionErrors) (*util.Set, error) {
+	// Parsing storageClassDeviceSets and parsing it to volume sources
+	c.spec.Storage.VolumeSources = append(c.spec.Storage.VolumeSources, c.prepareStorageClassDeviceSets(errs)...)
+
+	// QUESTION: at what point are volume sources added to the CephCluster CR? I need to make sure
+	// code that does that didn't get lost in this refactor
+	c.ValidStorage.VolumeSources = c.spec.Storage.VolumeSources
```

Comment:
@leseb ?

Reply:
The storageClassDeviceSets add the volume sources. I don't think we use the `c.spec.Storage.VolumeSources` though.

Reply:
What is the purpose of `c.spec.Storage.VolumeSources` to the user then? Is that to support OSDs on PVC via brownfield cases? If it doesn't have a use to the user, we might want to remove it from the spec.

Reply:
@jmolmo do you have any thoughts on this? `VolumeSources` seems to only be used internally and not exposed to users.